src/common/framework/test_group.ts

```diff
@@ -94,23 +96,44 @@ class TestGroup<F extends Fixture> implements TestGroupBuilder<F> {
   }
 }
 
-interface TestBuilderWithName<F extends Fixture, P extends {}> extends TestBuilderWithParams<F, P> {
+interface TestBuilderWithName<F extends Fixture>
+  extends TestBuilderWithParamsAndSubParams<F, {}, {}> {
   desc(description: string): this;
+  /** @deprecated use cases() and subcases() instead */
   params<NewP extends {}>(specs: Iterable<NewP>): TestBuilderWithParams<F, NewP>;
+  cases<NewP extends {}>(specs: Iterable<NewP>): TestBuilderWithParams<F, NewP>;
+  subcases<NewSubP extends {}>(
+    specs: (_: {}) => Iterable<NewSubP>
+  ): TestBuilderWithSubParams<F, NewSubP>;
 }
 
-interface TestBuilderWithParams<F extends Fixture, P extends {}> {
-  fn(fn: TestFn<F, P>): void;
+interface TestBuilderWithSubParams<F extends Fixture, SubP extends {}>
+  extends TestBuilderWithParamsAndSubParams<F, {}, SubP> {
+  /** @deprecated use cases() and subcases() instead */
+  params<NewP extends {}>(specs: Iterable<NewP>): TestBuilderWithParamsAndSubParams<F, NewP, SubP>;
+  cases<NewP extends {}>(specs: Iterable<NewP>): TestBuilderWithParamsAndSubParams<F, NewP, SubP>;
```

Comment:
Once .subcases() has been called, I don't think it makes sense to furthe call .cases(). So the hierarchy can be simplified a bit.